src/ucs/sys/socket.c

```diff
@@ -0,0 +1,214 @@
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_socket_connect_nb_get_status(int fd)
+{
+    socklen_t conn_status_sz;
+    int ret, conn_status;
+
+    conn_status_sz = sizeof(conn_status);
+
+    ret = getsockopt(fd, SOL_SOCKET, SO_ERROR,
+                     &conn_status, &conn_status_sz);
+    if (ret < 0) {
+        ucs_error("getsockopt(fd=%d) failed to get SOL_SOCKET(SO_ERROR): %m", fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if ((conn_status == EINPROGRESS) || (conn_status == EWOULDBLOCK)) {
+        return UCS_INPROGRESS;
+    }
+
+    if (conn_status != 0) {
+        ucs_error("SOL_SOCKET(SO_ERROR) status on fd %d: %s", fd, strerror(conn_status));
+        return UCS_ERR_UNREACHABLE;
+    }
+
+    return UCS_OK;
+}
+
+size_t ucs_sockaddr_sizeof(const struct sockaddr *addr, ucs_status_t *status_p)
+{
+    ucs_status_t status = UCS_OK;
+    size_t size;
+
+    switch (addr->sa_family) {
+    case AF_INET:
+        size = sizeof(struct sockaddr_in);
+        break;
+    case AF_INET6:
+        size = sizeof(struct sockaddr_in6);
+        break;
+    default:
+        ucs_error("unknown address family: %d", addr->sa_family);
+        status = UCS_ERR_INVALID_PARAM;
+        size = 0;
+        break;
+    }
+
+    if (status_p) {
+        *status_p = status;
+    }
+    return size;
+}
+
+int ucs_sockaddr_get_port(const struct sockaddr *addr)
+{
+    if (addr == NULL)
+        return 0;
+
+    switch (addr->sa_family) {
+    case AF_INET:
+        return ntohs(UCS_SIN_PORT(addr));
+    case AF_INET6:
+        return ntohs(UCS_SIN6_PORT(addr));
+    default:
+        ucs_error("unknown address family: %d", addr->sa_family);
+        return UCS_ERR_INVALID_PARAM;
+    }
+}
+
+const void *ucs_sockaddr_get_ipaddr(const struct sockaddr *addr)
+{
+    switch (addr->sa_family) {
+    case AF_INET:
+        return &UCS_SIN_ADDR(addr);
+    case AF_INET6:
+        return &UCS_SIN6_ADDR(addr);
+    default:
+        ucs_error("unknown address family: %d", addr->sa_family);
+        return NULL;
+    }
+}
+
+const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
+                             char *str, size_t max_size)
+{
+    static const char *err_msg = "Invalid address";
+    size_t str_len;
+    int ret;
+
+    if ((sock_addr->sa_family != AF_INET) && (sock_addr->sa_family != AF_INET6)) {
+        ucs_error("unknown address family: %d", sock_addr->sa_family);
+        return err_msg;
+    }
+
+    if (!inet_ntop(sock_addr->sa_family, ucs_sockaddr_get_ipaddr(sock_addr),
+                   str, max_size)) {
+        ucs_error("inet_ntop failed: %m");
+        return err_msg;
+    }
+
+    str_len = strlen(str);
+
+    ret = snprintf(str + str_len, max_size - str_len, ":%d",
+                   ucs_sockaddr_get_port(sock_addr));
+    if (ret < 0) {
+        ucs_error("snprintf failed: %d", ret);
+        return err_msg;
+    }
+
+    if (max_size > 0) {
+        str[ucs_min(max_size, (size_t)ret) + str_len] = '\0';
```

Comment:
> looks like `()` are wrong.

yes, it must be `ucs_min(max_size, (size_t)ret + str_len)`

> Also snprintf should null terminate itself

yes, right, but it handles error case (there is the test for this case)
this handles case when the size of passed buffer was smaller than required.
e.g.:
```
char buf[5];
const char *str = "192.168.0.1"; // 12 symbols (including null-terminated symbol)

snprintf(buf, 5, "%s", str); // it will fill 5 symbols of buf

for (i = 0; i < 5; i++) {
    printf("%c-", buf[i]);
}
// output:  "1-9-2-.-1-"

printf("%s", buf);
// output: "192.1<...>" <...> - some unexpected output until '\0' will be reached
```

Reply:
Thank you! Fixed the bug + 
1) added comment to make it clearer
2) removed unneeded check `max_size > 0` (i don't know why i added it there)